include/s3select_json_parser.h

```diff
@@ -0,0 +1,475 @@
+	  src_ = next_src_;
+	  begin_ = src_;
+	  size_ = next_size_;
+	  end_ = src_ + size_;
+
+	  next_src_ = 0;
+	  next_size_ = 0;
+	  return *src_;
+	}
+	else return 0;
+      }
+      return *src_++;
+    }
+
+    void resetBuffer(char* buff, size_t size)
+    {
+      if(!src_)
+      {//first time calling
+	begin_ = buff;
+	src_ = buff;
+	size_ = size;
+	end_= src_ + size_;
+	return;
+      }
+
+      if(!next_src_)
+      {//save the next-chunk that will be used upon parser reaches end of current buffer
+	next_src_ = buff;
+	next_size_ = size;
+      }
+      else
+      {// should not happen
+	std::cout << "can not replace pointers!!!" << std::endl;//TODO exception
+	return;
+      }
+    }
+
+    void saveRemainingBytes()
+    {//this routine called per each new chunk
+      //savine the remaining bytes, before its overriden by the next-chunk.
+      size_t copy_left_sz = getBytesLeft(); //should be very small
+      internal_buffer.assign(src_,copy_left_sz);
+      
+      src_ = internal_buffer.data();
+      begin_ = src_;
+      size_ = copy_left_sz;
+      end_= src_ + copy_left_sz;
+    }
+
+    size_t getBytesLeft() { return end_ - src_; }
+
+};
+
+class JsonParserHandler : public rapidjson::BaseReaderHandler<rapidjson::UTF8<>, JsonParserHandler> {
+
+  public:
+
+    typedef enum {OBJECT_STATE,ARRAY_STATE} en_json_elm_state_t;
+
+    typedef std::pair<std::vector<std::string>, Valuesax> json_key_value_t;
+
+    enum class row_state
+    {
+      NA,
+      ARRAY,
+      START_ROW,
+      OBJECT_START_ROW,
+      ARRAY_START_ROW,
+      ARRAY_END_ROW,
+      END_ROW
+    };
+
+    row_state state = row_state::NA;
+// TODO its possible there is no need to std::vector<std::string>, the json-idx is enough
+    std::function <int(JsonParserHandler::json_key_value_t&,int)> m_exact_match_cb;
+
+    std::vector <std::vector<std::string>> query_matrix{};
+    int row_count{};
+    std::vector <std::string> from_clause{};
+    bool prefix_match{};
+    Valuesax value;
+    ChunksStreamer stream_buffer;
+    bool init_buffer_stream;
+    rapidjson::Reader reader;
+    std::vector<en_json_elm_state_t> json_element_state;
+    std::string m_result;//debug purpose
+    std::vector<std::string> key_path;
+    std::function<int(void)> m_s3select_processing;
+    bool m_end_of_chunk;
+
+    JsonParserHandler() : init_buffer_stream(false),m_end_of_chunk(false)
+    {} 
+
+    std::string get_key_path()
+    {//for debug
+	  std::string res;
+	  for(const auto & i: key_path)
+	  {
+	    res.append(i);
+	    res.append(std::string("/"));
+	  }
+	  return res;
+    }
+
+    void dec_key_path()
+    {
+      if (json_element_state.size())  {
+        if(json_element_state.back() != ARRAY_STATE)  {
+	        if(key_path.size() != 0) {
+	          key_path.pop_back();
+          }
+        }
+      }
+      
+      if (key_path.size() < from_clause.size()) {
+        prefix_match = false;
+      } 
+      else if (prefix_match) {
```

Comment:
upon `dec_key_path` and prefix_match==true
its START_ARRAY_ROW (to validate) , thus its an element in ARRAY